README.md
# react-inner-image-zoom

[Demos](https://laurenashpole.github.io/react-inner-image-zoom)

![GitHub Actions][build-badge] [![npm package][npm-badge]][npm] [![TypeScript definitions on DefinitelyTyped][dt-badge]][dt]

A React component for magnifying an image within its original container. Zoom behavior can be triggered on click or hover and the zoomed image can be moved by dragging on touch devices and either dragging or pan on hover on non-touch devices. The component supports responsive images, loading placeholders, optional fullscreen zoom on mobile, and more.

## Installation

**Note:** Version 2.0.0 introduces React hooks and requires React v16.8.0 or above. To use this package with older versions of React, install with `npm install react-inner-image-zoom@1.3.0` or `yarn add react-inner-image-zoom@1.3.0` instead of the instructions below.

### NPM
```
npm install react-inner-image-zoom
```

### Yarn
```
yarn add react-inner-image-zoom
```

### TypeScript

For TypeScript users, type definitions are available through [DefinitelyTyped](https://definitelytyped.org/) and can be installed with:

```
npm install --save-dev @types/react-inner-image-zoom
```

### Styling

I was originally importing the CSS directly into the component but I've recently realized that makes too many assumptions about the wider build process. You can now download the raw CSS file at:

[/src/InnerImageZoom/styles.css](https://raw.githubusercontent.com/laurenashpole/react-inner-image-zoom/master/src/InnerImageZoom/styles.css)

or the minified raw minified version at:

[/src/InnerImageZoom/styles.min.css](https://raw.githubusercontent.com/laurenashpole/react-inner-image-zoom/master/src/InnerImageZoom/styles.min.css)

to include however you see fit. Or, if your setup supports it, import the files directory from your `node_modules` using:

```javascript
import 'react-inner-image-zoom/lib/InnerImageZoom/styles.css';
```

or:

```javascript
import 'react-inner-image-zoom/lib/InnerImageZoom/styles.min.css';
```

## Usage

Import and render the component:
```javascript
import InnerImageZoom from 'react-inner-image-zoom';

...

<InnerImageZoom src="/path/to/image.jpg" zoomSrc="/path/to/zoom-image.jpg" />
```

This is the simplest usage. For additional examples, visit the [demo page](https://laurenashpole.github.io/react-inner-image-zoom).


## Props

Prop | Type | Default | Description
--- | --- | --- | ---
src | String | | (Required) URL for the original image.
sources | Array | | A list of image sources for using the picture tag to serve the appropriate original image (see below for more details).
width | Number | | Width attribute for original image.
height | Number | | Height attribute for original image.
hasSpacer | Boolean | false | If true, gets the original image's aspect ratio based on the width and height props and creates a spacer to prevent cumulative layout shift.
imgAttributes | Object | | [Img](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attributes) and [global](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes) attributes for the original image (excluding `src`, `width`, `height`, and `style` which are set elsewhere). The imgAttributes keys should follow the [React DOM element](https://reactjs.org/docs/dom-elements.html) naming conventions.
zoomSrc | String | | URL for the larger zoom image. Falls back to original image src if not defined.
zoomScale | Number | 1 | Multiplied against the natural width and height of the zoomed image. This will generally be a decimal (example, 0.9 for 90%).
zoomPreload | Boolean | false | If set to true, preloads the zoom image instead of waiting for mouseenter and (unless on a touch device) persists the image on mouseleave.
moveType | String | pan | `pan` or `drag`. The user behavior for moving zoomed images on non-touch devices.
zoomType | String | click | `click` or `hover`. The user behavior for triggering zoom. When using `hover`, combine with `zoomPreload` to avoid flickering on rapid mouse movements.
fadeDuration | Number | 150 | Fade transition time in milliseconds. If zooming in on transparent images, set this to `0` for best results.
fullscreenOnMobile | Boolean | false | Enables fullscreen zoomed image on touch devices below a specified breakpoint.
mobileBreakpoint | Number | 640 | The maximum breakpoint for fullscreen zoom image when fullscreenOnMobile is true.
hideCloseButton | Boolean | false | Hides the close button on touch devices. If set to true, zoom out is triggered by tap.
hideHint | Boolean | false | Hides the magnifying glass hint.
className | String | | Custom classname for styling the component.
afterZoomIn | Function | | Function to be called after zoom in.
afterZoomOut | Function | | Function to be called after zoom out.

### Sources

This prop accepts an array of objects which it uses to create a picture tag and source elements. The component looks for the following optional properties and you can find additional details on responsive images [here](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images):

Prop | Type | Default | Description
--- | --- | --- | ---
srcSet | String | | Srcset attribute for source tag.
sizes | String | | Sizes attribute for source tag.
media | String | | An attribute containing a media condition for use with the srcset.
type | String | | An image MIME type. This is useful for using newer formats like WebP.

## Issues

Please submit issues or requests [here](https://github.com/laurenashpole/react-inner-image-zoom/issues).

Most of the implementation choices for this component are based on use cases I've encountered in the past. For example, I chose click to zoom as the default because it's been the most requested on product detail pages I've worked on. If there's a demand for additional triggers or other functionality, I'd be open to looking into it so feel free to ask. And if you want to talk through ideas first, check out the [discussions page](https://github.com/laurenashpole/react-inner-image-zoom/discussions).

If you're interested in contributing, check out the guidelines [here](https://github.com/laurenashpole/react-inner-image-zoom/blob/master/CONTRIBUTING.md).

## License

[MIT](https://github.com/laurenashpole/react-inner-image-zoom/blob/master/LICENSE)

[build-badge]: https://github.com/laurenashpole/react-inner-image-zoom/actions/workflows/release.yml/badge.svg

[npm-badge]: http://img.shields.io/npm/v/react-inner-image-zoom.svg?style=flat
[npm]: https://www.npmjs.com/package/react-inner-image-zoom

[dt-badge]: https://definitelytyped.org/badges/standard-flat.svg
[dt]: http://definitelytyped.org


CHANGELOG.md
# Changelog

## [3.0.2](https://github.com/laurenashpole/react-inner-image-zoom/compare/v3.0.1...v3.0.2) (2022-07-22)


### Fixed

- A bug re-zooming after clicking the close button on non-touch devices when `zoomPreload` is false.

🎉🎉🎉 Special thanks to [MaxDAyala](https://github.com/MaxdAyala) for tackling the following:

- A Firefox error when the zoomed image is dragged to the far left of the container.
- The timing of the fade out `visibility` and `opacity` transitions.
- An intermittent issue where zooming became disabled by panning in and out at a fast speed.

## [3.0.1](https://github.com/laurenashpole/react-inner-image-zoom/compare/v3.0.0...v3.0.1) (2022-06-12)

### Fixed

- Added `prop-types` to the `peerDependencies`.

## [3.0.0](https://github.com/laurenashpole/react-inner-image-zoom/compare/v2.1.0...v3.0.0) (2022-01-03)

### Changed

- Replaced `srcSet`, `sizes`, `alt`, and `title` props with `imgAttributes` to set the original image's attributes.
- Show close button when moveType is set to "drag" on all breakpoints.
- Switched from `setTimeout` to `onTransitionEnd` to check that zoomed image has finished fading out.

### Added

- This handy CHANGELOG.

### Fixed

- Added `stopPropagation` on touchmove to prevent events below fullscreen modal.

##  [2.1.0](https://github.com/laurenashpole/react-inner-image-zoom/compare/v2.0.3...v2.1.0) (2021-08-30)

### Added

- `title` prop to add attribute to original image.

##  [2.0.3](https://github.com/laurenashpole/react-inner-image-zoom/compare/v2.0.2...v2.0.3) (2021-08-05)

### Changed

- Use `touch-action` CSS property instead of `preventDefault` to prevent scroll on touchmove and drag.

### Fixed

- Sporadic missing zoom image in fullscreen modal caused by missing dimensions and incorrect positioning.

##  [2.0.2](https://github.com/laurenashpole/react-inner-image-zoom/compare/v2.0.1...v2.0.2) (2021-06-15)

### Fixed

- Incorrect initial zoom position in fullscreen modal.
- Persist the zoomed image after zoom out if `zoomPreload` is true.

##  [2.0.1](https://github.com/laurenashpole/react-inner-image-zoom/compare/v2.0.0...v2.0.1) (2021-03-12)

### Fixed

- Set the scaled image size based on `naturalWidth` and `naturalHeight` instead of `offsetWidth` and `offsetHeight`.

##  [2.0.0](https://github.com/laurenashpole/react-inner-image-zoom/compare/v1.3.0...v2.0.0) (2021-03-03)

### Changed

- Refactored using React hooks. All versions after 2.0.0 require React v16.8.0 or above.
- Renamed `startsActive` to `zoomPreload`

### Added

- `hideHint` prop to hide the magnifying glass icon.
- `hideCloseButton` prop to hide the close button on touch devices.
- `width`, `height`, and `hasSpacer` props to set the original image's width and height attributes and optionally generate a spacer based on those values to avoid cumulative layout shift. 
- CONTRIBUTING guide.
- ESLint and Prettier formatting.

##  [1.3.0](https://github.com/laurenashpole/react-inner-image-zoom/compare/v1.2.0...v1.3.0) (2020-11-24)

### Added

- `zoomScale` prop to set the size of the zoomed image.
- `startsActive` prop to load the zoomed image on render.

##  [1.2.0](https://github.com/laurenashpole/react-inner-image-zoom/compare/v1.1.1...v1.2.0) (2020-11-21)

### Added

- `zoomType` prop with "hover" option to trigger zoom on hover.

##  [1.1.1](https://github.com/laurenashpole/react-inner-image-zoom/compare/v1.1.0...v1.1.1) (2020-07-13)

### Fixed

- Removed unnecessary dragend events when image is not zoomed.

##  [1.1.0](https://github.com/laurenashpole/react-inner-image-zoom/compare/v1.0.6...v1.1.0) (2020-07-12)

### Added

- `moveType` prop with "drag" option for drag to move functionality on non-touch devices.

## [1.0.6](https://github.com/laurenashpole/react-inner-image-zoom/compare/v1.0.5...v1.0.6) (2020-05-22)

### Fixed

- Hide original image on zoom to support transparent zoom images.

## [1.0.5](https://github.com/laurenashpole/react-inner-image-zoom/compare/v1.0.0...v1.0.5) (2019-10-15)

### Changed

- Removed `styles.css` import from React component to allow for a greater variety of build approaches.

### Added

- Minified CSS file `styles.min.css`.
- "Styling" section in README file.

## [1.0.0](https://github.com/laurenashpole/react-inner-image-zoom/compare/e8e458231a32831a4332b4c009e7df2d68535ada...v1.0.0) (2019-06-19)

### Added

- InnerImageZoom React component.
- README and LICENSE.


pages/docs.js
import Layout from '../components/Layout';
import { Light as SyntaxHighlighter } from 'react-syntax-highlighter';
import js from 'react-syntax-highlighter/dist/cjs/languages/hljs/javascript';
import a11yLight from 'react-syntax-highlighter/dist/cjs/styles/hljs/a11y-light';
import styles from './Docs.module.css';

SyntaxHighlighter.registerLanguage('javascript', js);

const Docs = () => {
  return (
    <Layout title="Docs">
      <div className={styles.container}>
        <h2 className={styles.heading}>Docs</h2>

        <nav className={styles.nav} aria-label="Topics">
          <a href="#installation">Installation</a>
          <a href="#styling">Styling</a>
          <a href="#usage">Usage</a>
          <a href="#props">Props</a>
        </nav>

        <section className={styles.content}>
          <div>
            <h3 id="installation">Installation</h3>
            <p><b>Note:</b> Version 2.0.0 introduces React hooks and requires React v16.8.0 or above. To use this package with older versions of React, install with <code className={styles.inlineCode}>npm install react-inner-image-zoom@1.3.0</code> or <code className={styles.inlineCode}>yarn add react-inner-image-zoom@1.3.0</code> instead of the instructions below.</p>
            <h4>NPM</h4>
            <pre className={styles.code}><code>npm install react-inner-image-zoom</code></pre>
            <h4>Yarn</h4>
            <pre className={styles.code}><code>yarn add react-inner-image-zoom</code></pre>
            <h4>TypeScript</h4>
            <p>For TypeScript users, type definitions are available through DefinitelyTyped and can be installed with:</p>
            <pre className={styles.code}><code>npm install --save-dev @types/react-inner-image-zoom</code></pre>
          </div>

          <div>
            <h3 id="styling">Styling</h3>
            <h4>Download</h4>
            <p>Grab the raw CSS from Github to use however you see fit:</p>
            <p><a href="https://raw.githubusercontent.com/laurenashpole/react-inner-image-zoom/master/src/InnerImageZoom/styles.css" target="_blank" rel="noopener noreferrer">styles.css</a></p>
            <p>or the minified version:</p>
            <p><a href="https://raw.githubusercontent.com/laurenashpole/react-inner-image-zoom/master/src/InnerImageZoom/styles.min.css" target="_blank" rel="noopener noreferrer">styles.min.css</a></p>
            <h4>Import</h4>
            <p>If your setup supports it, import the CSS directly from your node_modules:</p>
            <SyntaxHighlighter language="javascript|html" style={a11yLight} customStyle={{ margin: 0, padding: '1.5rem', background: '#f2f3f7', borderRadius: '1px' }}>
              import 'react-inner-image-zoom/lib/InnerImageZoom/styles.css';
            </SyntaxHighlighter>
            <p>or the minified version:</p>
            <SyntaxHighlighter language="javascript|html" style={a11yLight} customStyle={{ margin: 0, padding: '1.5rem', background: '#f2f3f7', borderRadius: '1px' }}>
              import 'react-inner-image-zoom/lib/InnerImageZoom/styles.min.css';
            </SyntaxHighlighter>
          </div>

          <div>
            <h3 id="usage">Usage</h3>
            <p>Import and render the component:</p>
            <SyntaxHighlighter language="javascript|html" style={a11yLight} customStyle={{ margin: 0, padding: '1.5rem', background: '#f2f3f7', borderRadius: '1px' }}>
              {`import InnerImageZoom from 'react-inner-image-zoom';\n\n...\n\n<InnerImageZoom src="/path/to/image.jpg" zoomSrc="/path/to/zoom-image.jpg" />`}
            </SyntaxHighlighter>
          </div>

          <div>
            <h3 id="props">Props</h3>

            <ul className={styles.propsList}>
              <li><b>src (String): </b>Required. URL for the original image.</li>
              <li><b>sources (Array): </b>A list of image sources for using the picture tag to serve the appropriate original image (see below for more details).</li>
              <li><b>width (Number): </b>Width attribute for original image.</li>
              <li><b>height (Number): </b>Height attribute for original image.</li>
              <li><b>hasSpacer (Boolean): </b>Default false. If true, gets the original image's aspect ratio based on the width and height props and creates a spacer to prevent cumulative layout shift.</li>
              <li><b>imgAttributes (Object): </b><a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attributes" target="_blank" rel="noopener noreferrer">Img</a> and <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes" target="_blank" rel="noopener noreferrer">global</a> attributes for the original image (excluding <code className={styles.inlineCode}>src</code>, <code className={styles.inlineCode}>width</code>, <code className={styles.inlineCode}>height</code>, and <code className={styles.inlineCode}>style</code> which are set elsewhere). The imgAttributes keys should follow the <a href="https://reactjs.org/docs/dom-elements.html" target="_blank" rel="noopener noreferrer">React DOM element</a> naming conventions.</li>
              <li><b>zoomSrc (String): </b>URL for the larger zoom image. Falls back to original image src if not defined.</li>
              <li><b>zoomScale (Number): </b>Default 1. Multiplied against the natural width and height of the zoomed image. This will generally be a decimal (example, 0.9 for 90%).</li>
              <li><b>zoomPreload (Boolean): </b>Default false If set to true, preloads the zoom image instead of waiting for mouseenter and (unless on a touch device) persists the image on mouseleave.</li>
              <li><b>moveType (String): </b>Default pan. Accepts pan or drag options. The user behavior for moving zoomed images on non-touch devices.</li>
              <li><b>zoomType (String): </b>Default click. Accepts click or hover options. The user behavior for triggering zoom. When using hover, combine with zoomPreload to avoid flickering on rapid mouse movements.</li>
              <li><b>fadeDuration (Number): </b>Default 150. Fade transition time in milliseconds. If zooming in on transparent images, set this to 0 for best results.</li>
              <li><b>fullscreenOnMobile (Boolean): </b>Default false. Enables fullscreen zoomed image on touch devices below a specified breakpoint.</li>
              <li><b>mobileBreakpoint (Number): </b>Default 640. The maximum breakpoint for fullscreen zoom image when fullscreenOnMobile is true.</li>
              <li><b>hideCloseButton (Boolean): </b>Default false. Hides the close button on touch devices. If set to true, zoom out is triggered by tap.</li>
              <li><b>hideHint (Boolean): </b>Default false. Hides the magnifying glass hint.</li>
              <li><b>className (String): </b>Custom classname for styling the component.</li>
              <li><b>afterZoomIn (Function): </b>Function to be called after zoom in.</li>
              <li><b>afterZoomOut (Function): </b>Function to be called after zoom out.</li>
            </ul>

            <h4>Sources</h4>
            <p>This prop accepts an array of objects which it uses to create a picture tag and source elements. The component looks for the following optional properties and you can find additional details <a href="https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images">here</a>: </p>

            <ul className={styles.propsList}>
              <li><b>srcSet (String): </b>Srcset attribute for source tag.</li>
              <li><b>sizes (String): </b>Sizes attribute for source tag.</li>
              <li><b>media (String): </b>An attribute containing a media condition for use with the srcset.</li>
              <li><b>type (String): </b>An image MIME type. This is useful for using newer formats like WebP.</li>
            </ul>
          </div>
        </section>
      </div>
    </Layout>
  );
};

export default Docs;
